src/snovault/elasticsearch/indexer_state.py

```diff
@@ -0,0 +1,630 @@
+
+        if changed:  # alter notify even if error, so the same error doesn't flood log.
+            self.put_obj('notify', notify, 'default')
+
+    def display(self, uuids=None):
+        display = {}
+        display['state'] = self.get()
+        if display['state'].get('status','') == 'indexing' and 'cycle_started' in display['state']:
+            started = datetime.datetime.strptime(display['state']['cycle_started'],'%Y-%m-%dT%H:%M:%S.%f')
+            display['state']['indexing_elapsed'] = str(datetime.datetime.now() - started)
+        display['title'] = display['state'].get('title',self.state_id)
+        display['uuids in progress'] = self.get_count(self.todo_set)
+        display['uuids troubled'] = self.get_count(self.troubled_set)
+        display['uuids last cycle'] = self.get_count(self.last_set)
+        if self.followup_prep_list is not None:
+            display['to be handed off to other indexer(s)'] = self.get_count(self.followup_prep_list)
+        if self.title == 'primary':
+            for id in self.followup_lists:
+                display[id] = self.get_count(id)
+        else:
+            id = 'staged_for_%s_list' % (self.title)
+            display['staged by primary'] = self.get_count(id)
+
+        reindex = self.get_obj(self.override)
+        if reindex:
+            uuids = reindex.get('uuids')
+            if uuids is not None:
+                display['REINDEX requested'] = uuids
+            elif reindex.get('all_uuids',False):
+                display['REINDEX requested'] = 'all'
+        notify = self.get_notices()
+        if notify:
+            display['NOTIFY requested'] = notify
+        display['now'] = datetime.datetime.now().isoformat()
+
+        if uuids is not None:
+            uuids_to_show = []
+            uuid_list = self.get_obj(self.todo_set)
+            if not uuid_list:
+                uuids_to_show = 'No uuids indexing'
+            else:
+                uuid_start = 0
+                try:
+                    uuid_start = int(uuids)
+                except:
+                    pass
+                if uuid_start < uuid_list.get('count',0):
+                    uuid_end = uuid_start+100
+                    if uuid_start > 0:
+                        uuids_to_show.append("... skipped first %d uuids" % (uuid_start))
+                    uuids_to_show.extend(uuid_list['list'][uuid_start:uuid_end])
+                    if uuid_list.get('count',0) > uuid_end:
+                        uuids_to_show.append("another %d uuids..." % (uuid_list.get('count',0) - uuid_end))
+                elif uuid_start > 0:
+                    uuids_to_show.append("skipped past all %d uuids" % (uuid_list.get('count',0)))
+                else:
+                    uuids_to_show = 'No uuids indexing'
+            display['uuids in progress'] = uuids_to_show
+
+        return display
+
+@view_config(route_name='_indexer_state', request_method='GET', permission="index")
```

Comment:
This permission="index" is the same permission as
```
@view_config(route_name='index', request_method='POST', permission="index")
def index(request):
    ....
```
But this allows someone to request a reindexing.  Who exactly can do this.

Reply:
User's with index permissions... administrators.  Conceivably index permissions could be extended to non-administrators.  Though I can't imagine why.  I prefer this is restricted to index since that is THE thing the API does that we want to restrict access to.

Reply:
we should definitely check to be sure only administrators can do this!
@caseylitton permission=index was the permission set for the internal indexing calls.